impeller/docs/standalone_gles.md

```diff
@@ -0,0 +1,195 @@
+
+# Setup
+
+To get started with Impeller rendering, you need to setup a context and a renderer. For backend specific classes, the convention in Impeller is to append the backend name as a suffix. So if you need to create an instance of an `impeller::Context` for OpenGLES, look for `impeller::ContextGLES`.
+
+In our case, we need a `impeller::ContextGLES` and `impeller::Renderer`.
+
+## Setting up WSI with the EGL Toolkit
+
+Before we get to `OpenGLES` we need to do a bit of WSI. While we can resort to using EGL directly, Impeller has a handy toolkit for it.
+
+First create an EGL display connection:
+
+```c++
+egl::Display display;
+```
+
+Ask the display for a valid EGL configuration. Impeller needs an OpenGL ES 2.0 configuration.
+
+```c++
+egl::ConfigDescriptor egl_desc;
+egl_desc.api = egl::API::kOpenGLES2;
+egl_desc.samples = egl::Samples::kOne;
+egl_desc.color_format = egl::ColorFormat::kRGBA8888;
+egl_desc.stencil_bits = egl::StencilBits::kZero;
+egl_desc.depth_bits = egl::DepthBits::kZero;
+egl_desc.surface_type = egl::SurfaceType::kWindow;
+
+auto config = display.ChooseConfig(egl_desc);
+```
+
+Once a valid config has been obtained, create a context and window surface. Creating the window surface requires a native window handle. Get the appropriate one for your platform. For instance, on Android this is an `ANativeWindow`.
+
+```c++
+auto context = display.CreateContext(*config, /* sharegroup= */ nullptr );
+auto surface = display.CreateWindowSurface(*config, native_window_handle);
+```
+
+Now that we have a context, make it current on the calling thread. This will complete the setup of WSI.
+
+```c++
+context->MakeCurrent(*surface);
+```
+
+## Creating the OpenGL ES Context
+
+Impeller doesn't statically link against OpenGL ES. You need to give it a callback the returns the appropriate OpenGL ES function for given name. With EGL, this can be something as simple as:
+
+```c++
+auto resolver = [](const char* name) -> void* {
+  return reinterpret_cast<void*>(::eglGetProcAddress(name));
+};
+```
+
+Adjust as necessary.
+
+Once you have the resolver, you need to create an OpenGL ES proc table. The proc table contains the function pointers for the subset of the OpenGL ES API used by Impeller.
+
+```c++
+auto gl = std::make_unique<ProcTableGLES>(resolver);
+```
+
+Once the proc table is created, the resolver will no longer be invoked.
+
+Then you need to provide the context a shader library that contains a manifest of all the shaders the context will need at runtime. Remember, Impeller doesn't generate shaders at runtime. Instead `impellerc` generates blobs that can either be delivered out of band or be embedded directly in the binary. When embedding the blobs directly in the binary, find the symbols referring to the shader blob somewhere in the generated build artifacts. A vector of mappings to these blobs needs to be provided to create context creation factory.
```

Comment:
it's not immediately clear to me what "find the symbols" means here. Where would I look for them?

Reply:
This is very application specific. The engine embeds the blob directly in the binary but it can be loaded separately. In the WASM example, its fetched via an XHR. I'll add a separate document detailing this.